cpp/dolfin/fem/Form.cpp

```diff
@@ -232,3 +246,46 @@ void Form::set_vertex_domains(
   dP = vertex_domains;
 }
 //-----------------------------------------------------------------------------
+void Form::tabulate_tensor(
+    double* A, mesh::Cell cell,
+    Eigen::Ref<const EigenRowArrayXXd> coordinate_dofs) const
+{
+  // Switch integral based on domain from dx MeshFunction
+  std::uint32_t idx = 0;
+  if (dx)
+  {
+    // FIXME: range checks on cell and idx
+    idx = (*dx)[cell];
+  }
+
+  const bool* enabled_coefficients = _integrals.cell_enabled_coefficients(idx);
+  // Restrict coefficients to cell
+  for (std::size_t i = 0; i < _coefficients.size(); ++i)
+  {
+    if (!enabled_coefficients[i])
+      continue;
+    const auto coefficient = _coefficients.get(i);
+    const auto& element = _coefficients.element(i);
+    coefficient->restrict(_wpointer[i], element, cell, coordinate_dofs);
+  }
+
+  // Compute cell matrix
+  auto tab_fn = _integrals.cell_tabulate_tensor(idx);
+  tab_fn(A, _wpointer.data(), coordinate_dofs.data(), 1);
+}
+//-----------------------------------------------------------------------------
+void Form::initialise_w()
+{
+  const std::size_t num_coeffs = _coefficients.size();
+  std::vector<std::uint32_t> n = {0};
+  for (std::uint32_t i = 0; i < num_coeffs; ++i)
+  {
+    const auto& element = _coefficients.element(i);
```

Comment:
This all needs some comments. Could you not use `auto` here to make the type clear.